src/operator/control_flow.cc

```diff
@@ -480,6 +480,512 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+                                      const std::vector<NDArray>& _outputs) {
+  // inputs are dl / df(x)
+  // outputs are dl / dx
+  // where f is the current function,
+  // x is the input to the current function,
+  // TODO(Junru): avoid dynamic NDArray allocation
+  WhileLoopState &state = state_ptr.get_state<WhileLoopState>();
+  const WhileLoopParam& params = state.params;
+  // sanity checks
+  CHECK_EQ(_outputs.size() + 2U, (size_t) params.num_args);
+  CHECK_EQ(_outputs.size(), _req.size());
+  for (auto x : _req) {
+    CHECK_NE(x, kWriteInplace);
+  }
+  for (auto x : _outputs) {
+    CHECK_EQ(x.storage_type(), kDefaultStorage)
+          << "The while_loop operator doesn't support the sparse format";
+  }
+  std::vector<NDArray> outputs;
+  std::vector<OpReqType> req;
+  WhileLoopState::extract_by_loc(_outputs, params.func_input_locs, &outputs);
+  WhileLoopState::extract_by_loc(_req, params.func_input_locs, &req);
+  if (state.n_iterations == 0) {
+    for (int i = params.num_out_data; i < params.num_outputs; ++i) {
+      int j = params.func_var_locs[i - params.num_out_data];
+      mxnet::CopyFromTo(inputs[i], &outputs[j]);
+    }
+    state.Cleanup();
+    return;
+  }
+  // collect var_locs and out_locs, positions other than var_locs are out_locs, i.e.
+  // [0, var_locs[0])
+  // (var_locs[1], var_locs[2]),
+  // (var_locs[2], var_locs[3]),
+  // ...
+  // (var_locs[-2], var_locs[-1] = params.num_args - 2)
+  std::vector<dim_t> var_locs(params.func_var_locs.begin(), params.func_var_locs.end());
+  var_locs.push_back((dim_t) params.num_args - 2U);
+  sort(var_locs.begin(), var_locs.end());
+  // vectors for the backward loop
+  std::vector<NDArray> ograds(params.num_outputs);
+  std::vector<NDArray> igrads(outputs.size());
+  std::vector<OpReqType> iter_req(req.size());
+  for (int i = params.num_out_data; i < params.num_outputs; ++i)
+    ograds[i] = inputs[i];
+  const int n_iter = state.n_iterations;
+  for (int step = n_iter - 1; step >= 0; --step) {
+    // ograds[ : num_out_data] = inputs[ : num_out_data][step]
+    // ograds[num_out_data: ] is maintained in the end of each loop
+    std::transform(std::begin(inputs),
+                   std::begin(inputs) + params.num_out_data,
+                   std::begin(ograds),
+                   [step] (const NDArray &a) { return a.At(step); } );
+    // igrads[i] =
+    //    outputs[i]            (step == 0)
+    //    outputs[i]            (step != 0 && i not in loop_var_locs)
+    //    ArrayLike(outputs[i]) (step != 0 && i in loop_var_locs)
+    // iter_req =
+    //    kWriteTo              (step != 0           && i in loop_var_locs)
+    //    req[i]                (step == 0           && i in loop_var_locs)
+    //    kAddTo                (step != n_iters - 1 && i not in loop_var_locs)
+    //    req[i]                (step == n_iters - 1 && i not in loop_var_locs)
+    {
+      size_t i = 0;
+      for (size_t loc : var_locs) {
```

Comment:
what about the variables behind the loop variables? for example, the inputs of the loop are
[loop_var1, loop_var2, other_var1, other_var2]. How are other_var1 and other_var2 processed?

Reply:
In [line 745](https://github.com/junrushao1994/mxnet/blob/316b0f7a10dcaff6b7792aa2a68c279c934087f8/src/operator/control_flow.cc#L745), a dummy position is appended to `var_locs`